proposed/query.md

```diff
@@ -0,0 +1,173 @@
+interfaces/functionality first.
+
+[RFC 2119]: http://tools.ietf.org/html/rfc2119
+
+## 1. Specification
+
+### 1.1 Goal
+
+There are numerous ways of querying data from so many sources. Each system has its very own way of doing it. For instance, Doctrine ORM and Symfony Finder are both querying data with query builders (or alike), but there is no standardized interface to implement such data querying.
+
+The goal of this PSR it to define common interfaces to query data, no matter where they come from (arrays, iterators, etc.) or their format.
+
+### 1.2 Definitions
+
+*    **Query** - The object that will be used to query data. It receives a source as an `iterable`, which allows processing of arrays as well as other type of collections like generators.
+
+*    **Modifier** - A modifier is an action applied to the source, which will be executed before each operation. There can be multiple modifiers on a Query, for instance a `where` clause, an `order by`, a `limit` or even an `offset`. Operations are applied _after_ modifiers.
+
+*    **Operation** - An operation is a one-time action applied to a processed source, after all modifiers has been applied. An operation can return any type of data: generators, scalars, etc.
+
+*    **Context** - What holds additional information to be passed to the query, modifiers and operations in the case they need it.
+
+## 2. Interfaces
+
+### 2.1 QueryInterface
+
+Has the ability to query an iterable source.
+
+```php
+namespace Psr\Query;
+
+use Psr\Query\QueryContextInterface;
+use Psr\Query\QueryModifierInterface;
+use Psr\Query\QueryOperationInterface;
+
+/**
+ * Defines an object which will execute modifiers and operations to query data.
+ * A QueryInterface could receive any iterable depending on the implementation. This could go from the
+ * basic array of data to generators.
+ */
+interface QueryInterface
+{
+    /**namespace Psr\Query;
+     * Creates a new object to query data.
+     *
+     * @param iterable $source
+     *      The source to apply manipulations on.
+     *
+     * @param QueryContextInterface|null $context
+     *      The optional context that forward needed information to the Query for its execution.
+     *
+     * @return QueryInterface
+     */
+    public static function from(iterable $source, QueryContextInterface $context = null): QueryInterface;
+
+    /**
+     * Get the context of the current Query, possibly modified by the latter.
+     *
+     * @return QueryContextInterface
+     */
+    public function getContext(): QueryContextInterface;
+
+    /**
+     * Gets the source of the Query.
+     *
+     * @return iterable
+     */
+    public function getSource(): iterable;
+
+    /**
+     * Applies a modifier to the Query.
+     *
+     * @param QueryModifierInterface $modifier
+     *      The actual modifier to apply.
+     *
+     * @return QueryInterface
+     */
+    public function applyModifier(QueryModifierInterface $modifier): QueryInterface;
```

Comment:
Ibid.

Reply:
Fixed as well. And also learnt what "ibid" stands for, thanks! :smile: